walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,126 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<String> attendees = new ArrayList<String>(request.getAttendees());
+    long longDuration = request.getDuration();
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+
+    // Deal with special cases:
+    // Meeting is too long, impossible to ever hold it
+    if (longDuration > TimeRange.WHOLE_DAY.duration()) {
+      return possibleTimes;
+    }
+    // No events/no attendees, hold meeting anytime
+    if (events.size() == 0 || attendees.size() == 0) {
+      possibleTimes.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, TimeRange.WHOLE_DAY.duration()));
+      return possibleTimes;
+    }
+ 
+    // Since we know longDuration is less than one day, it can fit in an int
+    int duration = (int)longDuration;
+
+    // Filter events so that it only includes events that have one or more
+    // attendees who are coming to this event
+    ArrayList<Event> eventsByStart = new ArrayList<Event>();
+    for (Event event : events) {
+      HashSet<String> eventAttendees = new HashSet<String>(event.getAttendees());
+      HashSet<String> meetingAttendees = new HashSet<String>(attendees);
+      eventAttendees.retainAll(meetingAttendees); // set intersection
+      if (eventAttendees.size() > 0) {
+        eventsByStart.add(event);
+      }
+    }
+    Collections.sort(eventsByStart, Event.ORDER_BY_START_TIME);   
+
+    // Check again to see if there are any events
+    if (eventsByStart.size() == 0 ) {
+      possibleTimes.add(TimeRange.fromStartDuration(TimeRange.START_OF_DAY, TimeRange.WHOLE_DAY.duration()));
```

Comment:
You could put these two lines in a separate function since there are multiple conditions where it's true

Reply:
Good idea! I moved things around a bit and created a function in TimeRange to return a range over the whole day.

Reply:
Oh wait, TimeRange already has this! So I won't bother to re-create it.